service/evaluation/evaluation.go

```diff
@@ -0,0 +1,783 @@
+
+	// Get a map of the assessment results, so that we have all assessment results for a specific resource_id and metric_id together for evaluation
+	assessmentResultsMap := getAssessmentResultMap(assessmentResults)
+	for key, results := range assessmentResultsMap {
+		var (
+			nonCompliantAssessmentResults []string
+			status                        evaluation.EvaluationResult_EvaluationStatus
+		)
+
+		// If no assessment_results are available continue
+		if len(results) == 0 {
+			continue
+		}
+
+		// Otherwise, there are some results and first we assume that everything
+		// is compliant, unless someone proves it otherwise
+		status = evaluation.EvaluationResult_EVALUATION_STATUS_COMPLIANT
+
+		for i := range results {
+			if !results[i].Compliant {
+				nonCompliantAssessmentResults = append(nonCompliantAssessmentResults, results[i].GetId())
+				status = evaluation.EvaluationResult_EVALUATION_STATUS_NOT_COMPLIANT
+			}
+		}
+
+		// Create evaluation result
+		eval = &evaluation.EvaluationResult{
+			Id:                         uuid.NewString(),
+			Timestamp:                  timestamppb.Now(),
+			ControlCategoryName:        categoryName,
+			ControlId:                  controlId,
+			CloudServiceId:             toe.GetCloudServiceId(),
+			ControlCatalogId:           toe.GetCatalogId(),
+			ResourceId:                 key,
+			Status:                     status,
+			FailingAssessmentResultIds: nonCompliantAssessmentResults,
+		}
+
+		err = s.storage.Create(eval)
+		if err != nil {
+			log.Errorf("error storing evaluation result for resource ID '%s' and control ID '%s' in database: %v", results[0].GetResourceId(), controlId, err)
+			continue
+		}
+
+		log.Debugf("Evaluation result stored for ControlID '%s' and Cloud Service ID '%s' with ID '%s'.", controlId, toe.GetCloudServiceId(), eval.Id)
+	}
+
+	// If the parentSchedulerTag is not empty, we have do decrement the WaitGroup so that the parent control can also be evaluated when all sub-controls are evaluated.
+	if parentSchedulerTag != "" && s.wg[parentSchedulerTag] != nil {
+		s.wg[parentSchedulerTag].Done()
+	}
+}
+
+// getMetricIds returns the metric Ids for the given metrics
+func getMetricIds(metrics []*assessment.Metric) []string {
+	var metricIds []string
+
+	for _, m := range metrics {
+		metricIds = append(metricIds, m.GetId())
+	}
+
+	return metricIds
+}
+
+// getAllMetricsFromControl returns all metrics from a given controlId
+// For now a control has either sub-controls or metrics. If the control has sub-controls, get also all metrics from the sub-controls.
+func (s *Service) getAllMetricsFromControl(catalogId, categoryName, controlId string) (metrics []*assessment.Metric, err error) {
+	var subControlMetrics []*assessment.Metric
+
+	control, err := s.getControl(catalogId, categoryName, controlId)
```

Comment:
This seems to be very inefficient. You are issuing a RPC for each individual control inside each evaluation cycle. Wouldn't it make more sense to just fetch the complete catalog once (per evaluation cycle) and then use this information instead of requesting information per-control?

Reply:
That's a good point.